oauth2/handler.go

```diff
@@ -689,6 +712,167 @@ func (h *Handler) getOidcUserInfo(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// swagger:route GET /oauth2/device/verify oAuth2 performOAuth2DeviceVerificationFlow
+//
+// # OAuth 2.0 Device Verification Endpoint
+//
+// This is the device user verification endpoint. The user is redirected here when trying to login using the device flow.
+//
+//	Consumes:
+//	- application/x-www-form-urlencoded
+//
+//	Schemes: http, https
+//
+//	Responses:
+//	  302: emptyResponse
+//	  default: errorOAuth2
+func (h *Handler) performOAuth2DeviceVerificationFlow(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	ctx := r.Context()
+
+	consentSession, f, err := h.r.ConsentStrategy().HandleOAuth2DeviceAuthorizationRequest(ctx, w, r)
+	if errors.Is(err, consent.ErrAbortOAuth2Request) {
+		x.LogAudit(r, nil, h.r.AuditLogger())
+		return
+	}
+
+	if e := &(fosite.RFC6749Error{}); errors.As(err, &e) {
+		x.LogAudit(r, err, h.r.AuditLogger())
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	if err != nil {
+		x.LogError(r, err, h.r.Logger())
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	req, err := h.r.OAuth2Storage().GetDeviceCodeSessionByRequestID(ctx, f.DeviceCodeRequestID.String(), &Session{})
+	if err != nil {
+		x.LogError(r, err, h.r.Logger())
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+	req.SetUserCodeState(fosite.UserCodeAccepted)
+	session, err := h.updateSessionWithRequest(ctx, consentSession, f, r, req, req.GetSession().(*Session))
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+	req.SetSession(session)
+	// Update the device code session with
+	//   - the claims for which the user gave consent
+	//   - the granted scopes
+	//   - the granted audiences
+	//   - the user_code_state set to accepted
+	// This marks it as ready to be used for the token exchange endpoint.
+	err = h.r.OAuth2Storage().UpdateDeviceCodeSessionByRequestID(ctx, f.DeviceCodeRequestID.String(), req)
```

Comment:
This is a bit awkward. When the user successfully logs in, there are 3 writes happening to the database all by this function. First the flow is written in the database (in [consent/strategy_default.go](https://github.com/ory/hydra/pull/3851/files#diff-265e72885dd37a9d8c6f81e2c928b2bfad035cecd98051ef2f04ea88a56ed540R1245)), the device auth request is updated and then the oidc session is updated.

I would like to have them grouped in a transaction. Currently there is the edge case where:
- 2 browser tabs open the device verification URL and input the same code
- Each user agent will get a different flow
- If both flows are completed, the one that got there first will see the device complete page and the other one will get an error

This will result in both flows being persisted to the database. If we were to wrap the flow creation and the device code update in a single transaction we would solve this issue, but the current code structure makes it difficult to implement (I don't want to start a transaction every time `HandleOAuth2DeviceAuthorizationRequest` is called, as most of the time nothing will be written to the database)

Reply:
I see, so it is mainly hard because you cannot tell whether the stateless flow will be updated (no DB write), or the final flow will be updated in the DB, correct?
The consequence of the flow being written is not security-relevant, right? It's rather a UX improvement? I think in that case we can fix that later.

Reply:
No I don't think it has any security implications. I will just add a TODO on the code so that this does not get lost